### ISSUE TITLE
countTotalDeps.js for dev deps.

### ISSUE THREAD

**Issue by jdalton** [anvaka/npmrank#7]
It would be neat to have the same regexp search for dev deps too.

**Comment by anvaka** [comment]
You can pass optional argument at the end to specify which graph file should be used. E.g.

```
node countTotalDeps.js "^lodash$" ./data/devDependencies.out.graph
```

**Comment by jdalton** [comment]
Ahh snaaoo! Thanks!

**Comment by jdalton** [comment]
I get weird results:

```bash
node countTotalDeps.js "^lodash$" ./data/devDependencies.out.graph
```
```bash
> 97066 unique packages depend on ^lodash$ out of total 178872 packages
That is 54.266%
```

**Comment by anvaka** [comment]
The number is too large?

Note: You are using devDependencies in this case. 

Let me add an option to actually print those dependents, so that we can verify why it's so high

**Comment by jdalton** [comment]
Ya it seems too high. The results are similar if you do "^typescript$" which I used to sanity check.

**Comment by anvaka** [comment]
TL;DR lodash, jquery and many other depends on TypeScript through devDependencies. 

See the proof below (it's not the shortest one, but accurate). You can check yourself by running:

```
node countTotalDeps.js "^typescript$" ./data/devDependencies.out.graph --print-names
```

It will print a tree of dependents.

lodash [devDepends on jquery](https://github.com/lodash/lodash/blob/9121ab95eaf45e20bf28b2e5dc53e1ecaf638a2e/package.json#L9)
jquery [devDepends on commitplease](https://github.com/jquery/jquery/blob/669cb16d763cb5486dadd56ec15a17b2b0303571/package.json#L28)
commitplease [devDepends on jscs](https://github.com/jzaefferer/commitplease/blob/d7efaf512450347dc47f2c6f15c6b32a1e5ee163/package.json#L28)
jscs [devDepends on xml2js](https://github.com/jscs-dev/node-jscs/blob/3da8359f4cb76467da7c23964807a2a55efde6f2/package.json#L98)
xml2js [devDepends on diff](https://github.com/Leonidas-from-XIV/node-xml2js/blob/43235366b0961ed86b9f3308ed59ef1c2f334c55/package.json#L72)
diff [devDepends on grunt-karma](https://github.com/kpdecker/jsdiff/blob/06ba84588d3398ddc5855129bdc7649bd64e0ac3/package.json#L46)
grunt-karma [devDepends on eslint-plugin-react](https://github.com/karma-runner/grunt-karma/blob/d2f24ae1d09e47b3c9c4af4bf6036787d65cca0b/package.json#L31)
eslint-plugin-react [devDepends on babel-eslint](https://github.com/yannickcr/eslint-plugin-react/blob/675ab31cc046c28597c4d757baca19d12e851da0/package.json#L26)
babel-eslint [devDepends on espree](https://github.com/babel/babel-eslint/blob/0db61dfa9279669fcc7e689a6d43a78ba91e2647/package.json#L30)
espree [devDepends on leche](https://github.com/eslint/espree/blob/da58617f378c17fb01273f374f439956fb465aa7/package.json#L48)
leche [devDepends on jsdoc](https://github.com/box/leche/blob/615f4ca06a010a230925d31b32f1bd25701d9099/package.json#L35)
jsdoc [devDepends on tv4](https://github.com/jsdoc3/jsdoc/blob/93bd06c9e24e81bd9057845ca0102edafd270500/package.json#L34)
tv4 [devDepends on grunt-markdown](https://github.com/geraintluff/tv4/blob/550ffed36db90fa8a4975a2cfdc84f3f3bac8e9e/package.json#L49)
grunt-markdown [devDepends on cheerio](https://github.com/treasonx/grunt-markdown/blob/ef4a8fd7a4f0335e27e1abccaf61f44b56b253d9/package.json#L46)
cheerio [devDepends on jsdom](https://github.com/cheeriojs/cheerio/blob/29d38e32889417390ee682b4169d6f9eee5f9d14/package.json#L35)
jsdom [devDepends on wd](https://github.com/tmpvar/jsdom/blob/b965621bcb94aafa22d232ea2598380dcb5d4bdb/package.json#L49)
wd [devDepends on http-proxy](https://github.com/admc/wd/blob/4543ab8180b2cb5bba0416c63c71730d0c72e424/package.json#L64)
http-proxy [devDepends on socket.io](https://github.com/nodejitsu/node-http-proxy/blob/30e3b371de0116e40e15156394f31c7e0b0aa9f1/package.json#L28)
socket.io [devDepends on superagent](https://github.com/socketio/socket.io/blob/ae5420b7275c61c87805e8e0ffa3698469fd5be5/package.json#L35)
superagent [devDepends on Base64](https://github.com/visionmedia/superagent/blob/2f5a7f5da5be16cc61a735dc0750086404184f0b/package.json#L47)
Base64 [devDepends on uglify-js](https://github.com/davidchambers/Base64.js/blob/b36932896c95d60db684f92e14468f233239bc25/package.json)
uglify-js [devDepends on acorn](https://github.com/mishoo/UglifyJS2/blob/881bda7f59cd62f1123fafa2da817b8dc618c19f/package.json#L38)
acorn [devDepends on browserify](https://github.com/marijnh/acorn/blob/b12b0b3c5affd944037d59b5eeb824abb4cf7893/package.json#L37)
browserify [devDepends on backbone](https://github.com/substack/node-browserify/blob/34e46a97706f35221dd5654e4240d20c608c1778/package.json#L74)
backbone [devDepends on karma](https://github.com/jashkenas/backbone/blob/7db3e5e627ae3c2c3268ce1f91d428d4c099de47/package.json#L13)
karma [devDepends on supertest-as-promised](https://github.com/karma-runner/karma/blob/16eb7a48271a8915abee73e720447a324268e506/package.json#L296)
supertest-as-promised [devDepends on when](https://github.com/WhoopInc/supertest-as-promised/blob/e3491b024c695ac026fde28b9941eb0b51de6633/package.json#L39)
when [devDepends on rest](https://github.com/cujojs/when/blob/9d120fac78c09f6eeb62baa16f36c1297e13d6cb/package.json#L55)
rest [devDepends on wire](https://github.com/cujojs/rest/blob/c213b37dde90413a52b8df3eef6fc4dd57520b9c/package.json#L37)
wire [devDepends on bower](https://github.com/cujojs/wire/blob/bb3772eab2809aaccac71025fef2b0b8cd2326ae/package.json#L46)
bower [devDepends on nock](https://github.com/bower/bower/blob/cb019c405b18dcee20aeb4370b6d3d10b9377e4c/package.json#L73)
nock [devDepends on aws-sdk](https://github.com/pgte/nock/blob/b73b6032b82714c7955c648c0f702d80146ddc97/package.json#L156)
aws-sdk [devDepends on eslint](https://github.com/aws/aws-sdk-js/blob/34c956b2543d5d26f73ca03c006e051a8cd1d4fe/package.json#L19)
eslint [devDepends on beefy](https://github.com/eslint/eslint/blob/5a599c70a98eaed0d43a8f28690921ef4ad0702d/package.json#L70)
beefy [devDepends on mock](https://github.com/chrisdickinson/beefy/blob/cc4bfd243897b4cfb47e91e06841e529a15b0c66/package.json#L48)
mock [devDepends on testem](https://github.com/Colingo/mock/blob/8da6f76728012deb6f0485209d3a3ca37e9c2780/package.json#L23)
testem [devDepends on dirty-chai](https://github.com/airportyh/testem/blob/71fb40a1bcd1da36b37de5677a9423d143d70c98/package.json#L67)
dirty-chai [devDepends on chai-as-promised](https://github.com/prodatakey/dirty-chai/blob/ff3248c9673ee1b401438593a19b451e119dbd4b/package.json#L25)
chai-as-promised [devDepends on ecstatic](https://github.com/domenic/chai-as-promised/blob/5f20e6c8d2297cb0287b71704ae0366492ce1a01/package.json#L35)
ecstatic [devDepends on express](https://github.com/jfhbrook/node-ecstatic/blob/01510b56c75442346012cd87ef01127a6831df5e/package.json#L32)
express [devDepends on jade](https://github.com/strongloop/express/blob/ef7ad681b245fba023843ce94f6bcb8e275bbb8e/package.json#L69)
jade [devDepends on less](https://github.com/jadejs/jade/blob/ba7c32b98af6ab91ca1b416c20696825aa01e66a/package.json#L57)
less [devDepends on jit-grunt](https://github.com/less/less.js/blob/59c012c16a71eca2fd9f354b8db8fffaa5e14d6d/package.json#L60)
jit-grunt [devDepends on intelli-espower-loader](https://github.com/shootaroo/jit-grunt/blob/7a172eee44bffae0f40ab76758aafc9f52238ff2/package.json#L41)
intelli-espower-loader [devDepends on espower-loader](https://github.com/power-assert-js/intelli-espower-loader/blob/69a63bdf3bd86cc0a4bbe4776dfe342c4d9ee30d/package.json#L24)
espower-loader [devDepends on empower](https://github.com/power-assert-js/espower-loader/blob/88249db959fd52c90fb2a45c84b39f09e027ee76/package.json#L21)
empower [devDepends on blanket](https://github.com/power-assert-js/empower/blob/c788718db868e3a5927cf166d08b2ef9336f9d0e/package.json#L24)
blanket [devDepends on async](https://github.com/alex-seville/blanket/blob/e26d3dc94aaa9183c7d669ea0c80d97450fc5213/package.json#L34)
async [devDepends on bluebird](https://github.com/caolan/async/blob/92f78aebad222d60c13e4299c0e723f2fe2d6611/package.json)
bluebird [devDepends on kefir](https://github.com/petkaantonov/bluebird/blob/386ba4f7d588693e5d675290a6b7fade08e0d626/package.json#L67)
kefir [devDepends on flyd](https://github.com/rpominov/kefir/blob/858c7a9810bc834d2f343129617c3b99e1bcd081/package.json#L38)
flyd [devDepends on transducers.js](https://github.com/paldepind/flyd/blob/aeb68452a65d905b6c7c41ccaffbb0d2c8ea599a/package.json#L21)
transducers.js [devDepends on immutable](https://github.com/jlongster/transducers.js/blob/7615a460d52e34ba73a61647bf69887a07a00fa3/package.json#L30)
immutable [devDepends on typescript](https://github.com/facebook/immutable-js/blob/e768a54393ca83ad0374ebf2a329a5a4c6bc4501/package.json#L51)

So, yes. Developers love devDependencies and use various tools :).

**Comment by jdalton** [comment]
![](https://cloud.githubusercontent.com/assets/4303/9650948/ae78caf8-51be-11e5-97e4-f79abe9586cc.gif)

**Comment by jdalton** [comment]
Still weird lots and lots of packages report ~54%.

**Comment by jdalton** [comment]
@anvaka  How does the deps of deps reporting work for dev deps. Lodash's is a dev dep for ~1,400 packages. So is the total 97,066 count based on the deps and deps deps of the 1,400 or the dev deps and dev deps dev deps of the 1,400?

**Comment by anvaka** [comment]
It's only devDeps all the way down.

**Comment by jdalton** [comment]
I'm not sure it meshes right. As dev deps of devDeps don't really flow. Once you start digging into deps of devDeps its really just deps (not dev deps).

**Comment by anvaka** [comment]
Can you give an example of such path? It would help me to debug this.

**Comment by jdalton** [comment]
Something like the query is trying to answer what kind of impact does pkg xyz have in the realm of dev deps. If pkg xyz was to break what's the potential effect. The results don't really align with that.

**Comment by anvaka** [comment]
I think I misunderstood you. Are you suggesting to to implement utility which would measure "devDeps" impact of a package "foo" as number of packages that dev-depend on "foo", plus number of transitive regular dependents of those packages?

**Comment by jdalton** [comment]
Maybe that's goofy. I'll kick it around a bit.
